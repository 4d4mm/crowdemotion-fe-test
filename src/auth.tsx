
const TOKEN_KEY = 'token';

export const login = (token:string) => {
    localStorage.setItem(TOKEN_KEY, token)
}

export const logout = () => {
    localStorage.removeItem(TOKEN_KEY)
}

export const getToken = () => localStorage.getItem(TOKEN_KEY)

export const isLoggedIn = () => {
    const token = localStorage.getItem(TOKEN_KEY);
    if (!token) {
        return false
    }

    const data = token.split('.')[1];
    const decodedData = atob(data);
    const jsonData = JSON.parse(decodedData);
    const expires = jsonData.exp;

    if (expires*1000 < (new Date()).getTime()) {
        return false;
    }

    return true
}
