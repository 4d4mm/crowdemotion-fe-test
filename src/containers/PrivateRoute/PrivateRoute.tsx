import * as React from "react";
import { Route, RouteComponentProps, RouteProps, Redirect } from "react-router";
import { isLoggedIn } from '../../auth'

const Unauthenticated = () => <Redirect to="/login" />

interface PrivateRouteProps extends RouteProps {
  unauthenticated?: React.ComponentType<RouteComponentProps<any>> | React.ComponentType<any>;
}

const PrivateRoute: React.SFC<PrivateRouteProps> = ({
  component: Component,
  unauthenticated: UnauthenticatedComponent = Unauthenticated,
  ...rest
}) => {
  if (!Component) {
    return null;
  }
  return (
    <Route
      {...rest}
      render={(props: RouteComponentProps<{}>) =>
        isLoggedIn() ? <Component {...props} /> : <UnauthenticatedComponent  {...props} />
      }
    />
  );
};

export default PrivateRoute;
