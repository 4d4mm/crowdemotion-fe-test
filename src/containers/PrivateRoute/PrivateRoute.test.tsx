import React from "react";
import { MemoryRouter, Route } from "react-router-dom";
import { mount } from "enzyme";
import PrivateRoute from "./PrivateRoute";
import { login } from '../../auth';

const DummyComponent = () => (<div>DummyComponent</div>)
describe("PrivateRoute", () => {
  it("renders with default ui", () => {
    const renderedProvateRouteNoComponent = mount(
        <MemoryRouter initialEntries={[{pathname:'/', key:'test'}]}>
            <PrivateRoute path="/" key="test" component={DummyComponent} unauthenticated={DummyComponent}/>
        </MemoryRouter>
    );
    expect(renderedProvateRouteNoComponent).toMatchSnapshot();
  });
  it("renders null if the component is not defined", () => {
    login('testtoken')
    const renderedProvateRouteNoComponent = mount(
        <MemoryRouter initialEntries={[{pathname:'/', key:'test'}]}>
            <PrivateRoute path="/" key="test" unauthenticated={DummyComponent}/>
        </MemoryRouter>
    );
    expect(renderedProvateRouteNoComponent.find('PrivateRoute').children().length).toBe(0);
  });
});
