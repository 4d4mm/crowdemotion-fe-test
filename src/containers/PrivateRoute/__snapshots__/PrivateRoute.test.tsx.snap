// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRoute renders with default ui 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "test",
        "pathname": "/",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "test",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "test",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <PrivateRoute
      component={[Function]}
      key="test"
      path="/"
      unauthenticated={[Function]}
    >
      <Route
        path="/"
        render={[Function]}
      >
        <DummyComponent
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "canGo": [Function],
              "createHref": [Function],
              "entries": Array [
                Object {
                  "hash": "",
                  "key": "test",
                  "pathname": "/",
                  "search": "",
                },
              ],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "index": 0,
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "key": "test",
                "pathname": "/",
                "search": "",
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "key": "test",
              "pathname": "/",
              "search": "",
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
        >
          <div>
            DummyComponent
          </div>
        </DummyComponent>
      </Route>
    </PrivateRoute>
  </Router>
</MemoryRouter>
`;
